File: ordermanager-backend/backtestapp_bak/examples/check_mike_v1_same.py
#!/usr/bin/env python3
"""Compare mike_v1 exported debug extrema vs offline detector recomputation."""

from __future__ import annotations

import argparse
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from pmbacktest.strategies.mike_v1 import _SideDetector

WINDOW_MS = 5 * 60 * 1000


@dataclass(frozen=True)
class EventRow:
    ts_ms: int
    kind: str
    side: str


def _fnum(value: object) -> float | None:
    if isinstance(value, (int, float)):
        v = float(value)
        if v == v and v not in (float("inf"), float("-inf")):
            return v
    return None


def _pnum(params: dict[str, object], *keys: str, default: float) -> float:
    for k in keys:
        v = _fnum(params.get(k))
        if v is not None:
            return v
    return float(default)


def _pstr(params: dict[str, object], *keys: str, default: str) -> str:
    for k in keys:
        v = params.get(k)
        if isinstance(v, str) and v.strip():
            return v.strip()
    return default


def _extract_saved_events(rows: list[dict[str, object]]) -> list[EventRow]:
    out: list[EventRow] = []
    for row in rows:
        if row.get("phase") != "accepted_extrema":
            continue
        kind = row.get("kind")
        side = row.get("side")
        if not isinstance(kind, str) or not isinstance(side, str):
            continue
        ts = _fnum(row.get("emitted_ts_ms"))
        if ts is None:
            ts = _fnum(row.get("ts_ms"))
        if ts is None:
            continue
        out.append(EventRow(ts_ms=int(ts), kind=kind, side=side))
    out.sort(key=lambda x: (x.ts_ms, x.side, x.kind))
    return out


def _make_detector(params: dict[str, object]) -> _SideDetector:
    location_mode = _pstr(params, "extrema_location_mode", "extremaLocationMode", default="spread_anchored").lower()
    return _SideDetector(
        span=int(_pnum(params, "span", default=20)),
        noise_filter=_pstr(params, "noise_filter", "noiseFilter", default="gaussian").lower(),
        structural_hysteresis=_pnum(
            params,
            "quote_structural_hysteresis_cents",
            "quoteStructuralHysteresisCents",
            default=0.1,
        ),
        sustained_tangent_ms=int(_pnum(params, "sustained_tangent_ms", "sustainedTangentMs", default=1000)),
        max_valley_ask_exclusive=_pnum(
            params,
            "valley_max_raw_ask_cents",
            "valleyMaxRawAskCents",
            default=45.0,
        ),
        valley_to_peak_min_cents=_pnum(
            params,
            "valley_to_peak_raw_spread_min_cents",
            "valleyToPeakRawSpreadMinCents",
            default=10.0,
        ),
        anchor_to_quote_series=location_mode == "spread_anchored",
    )


def _slice_by_window(events: list[EventRow], start_ms: int | None, rounds: int) -> list[EventRow]:
    if start_ms is None:
        return events
    end_ms = start_ms + max(1, rounds) * WINDOW_MS
    return [x for x in events if start_ms <= x.ts_ms < end_ms]


def _recompute_events(result: dict[str, Any]) -> list[EventRow]:
    meta = result.get("meta") if isinstance(result.get("meta"), dict) else {}
    params = meta.get("strategy_params") if isinstance(meta, dict) and isinstance(meta.get("strategy_params"), dict) else {}
    equity = result.get("equity")
    if not isinstance(equity, list):
        raise ValueError("result JSON has no `equity` array")
    points = [x for x in equity if isinstance(x, dict)]
    points.sort(key=lambda x: int(_fnum(x.get("timestamp_ms")) or -1))

    yes = _make_detector(params)
    no = _make_detector(params)
    out: list[EventRow] = []
    round_start_ms: int | None = None

    for p in points:
        ts = _fnum(p.get("timestamp_ms"))
        if ts is None:
            continue
        ts_i = int(ts)
        this_round_start = ts_i - (ts_i % WINDOW_MS)
        if round_start_ms != this_round_start:
            if round_start_ms is not None:
                yes = _make_detector(params)
                no = _make_detector(params)
            round_start_ms = this_round_start

        yes_ask = _fnum(p.get("yes_ask"))
        yes_bid = _fnum(p.get("yes_bid"))
        no_ask = _fnum(p.get("no_ask"))
        no_bid = _fnum(p.get("no_bid"))
        if yes_ask is None:
            yes_ask = _fnum(p.get("yes"))
        if no_ask is None:
            no_ask = _fnum(p.get("no"))
        if yes_bid is None:
            yes_bid = yes_ask
        if no_bid is None:
            no_bid = no_ask
        if yes_ask is None or yes_bid is None or no_ask is None or no_bid is None:
            continue

        sig_yes = yes.update(ts_ms=ts_i, ask=yes_ask, bid=yes_bid)
        if sig_yes and yes.last_emitted_idx is not None:
            out.append(EventRow(ts_ms=int(yes.times[yes.last_emitted_idx]), kind=sig_yes, side="yes"))
        sig_no = no.update(ts_ms=ts_i, ask=no_ask, bid=no_bid)
        if sig_no and no.last_emitted_idx is not None:
            out.append(EventRow(ts_ms=int(no.times[no.last_emitted_idx]), kind=sig_no, side="no"))

    out.sort(key=lambda x: (x.ts_ms, x.side, x.kind))
    return out


def _print_diff(left: list[EventRow], right: list[EventRow]) -> int:
    i = 0
    while i < len(left) and i < len(right):
        if left[i] != right[i]:
            print("Mismatch at index", i)
            print(" saved :", left[i])
            print(" recompute:", right[i])
            return 1
        i += 1
    if len(left) != len(right):
        print("Length mismatch:")
        print(" saved :", len(left))
        print(" recompute:", len(right))
        if i < len(left):
            print(" next saved :", left[i])
        if i < len(right):
            print(" next recompute:", right[i])
        return 1
    print("OK: saved events and recomputed events match exactly.")
    print("event_count =", len(left))
    return 0


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("result_json", type=Path, help="Path to exported backtest result JSON")
    parser.add_argument(
        "--round-start-ms",
        type=int,
        default=None,
        help="Optional UTC 5m window start timestamp; compare only this window (or windows with --rounds).",
    )
    parser.add_argument("--rounds", type=int, default=1, help="Window count when --round-start-ms is provided")
    args = parser.parse_args()

    obj = json.loads(args.result_json.read_text(encoding="utf-8"))
    if not isinstance(obj, dict):
        raise ValueError("Result JSON root must be an object")

    raw_debug = obj.get("strategy_debug_events")
    if not isinstance(raw_debug, list):
        raise ValueError(
            "No `strategy_debug_events` found. Re-run mike_v1 with strategy param `debug_trace=true`."
        )
    saved = _extract_saved_events([x for x in raw_debug if isinstance(x, dict)])
    recomputed = _recompute_events(obj)
    saved = _slice_by_window(saved, args.round_start_ms, args.rounds)
    recomputed = _slice_by_window(recomputed, args.round_start_ms, args.rounds)
    return _print_diff(saved, recomputed)


if __name__ == "__main__":
    raise SystemExit(main())

File: ordermanager-backend/backtestapp/web/scripts/checkMikeV1Same.ts
#!/usr/bin/env -S npx tsx

import fs from "node:fs";
import path from "node:path";

import { filterAlternatingAskSwingCents } from "../src/utils/alternatingAskExtremaFilter";
import { discreteTangentExtremaIndices } from "../src/utils/discreteTangentExtrema";
import { applyNoiseFilter } from "../src/utils/noiseFilters";

type AnyObj = Record<string, unknown>;

type EventRow = {
  ts_ms: number;
  side: "yes" | "no";
  kind: "valley" | "peak";
};

const WINDOW_MS = 5 * 60 * 1000;

function fnum(v: unknown): number | null {
  return typeof v === "number" && Number.isFinite(v) ? v : null;
}

function parseArgs(argv: string[]): { jsonPath: string; roundStartMs: number | null; rounds: number } {
  if (argv.length < 3) {
    throw new Error(
      "Usage: npm run check:mike-v1-same -- <result.json> [--round-start-ms <ms>] [--rounds <n>]",
    );
  }
  const out = { jsonPath: argv[2]!, roundStartMs: null as number | null, rounds: 1 };
  for (let i = 3; i < argv.length; i++) {
    const a = argv[i]!;
    if (a === "--round-start-ms") {
      const v = Number(argv[++i] ?? "");
      if (!Number.isFinite(v)) throw new Error("--round-start-ms requires a finite number");
      out.roundStartMs = Math.trunc(v);
      continue;
    }
    if (a === "--rounds") {
      const v = Number(argv[++i] ?? "");
      if (!Number.isFinite(v) || v < 1) throw new Error("--rounds requires integer >= 1");
      out.rounds = Math.trunc(v);
      continue;
    }
    throw new Error(`Unknown arg: ${a}`);
  }
  return out;
}

function pnum(params: AnyObj, keys: string[], fallback: number): number {
  for (const k of keys) {
    const v = fnum(params[k]);
    if (v != null) return v;
  }
  return fallback;
}

function pstr(params: AnyObj, keys: string[], fallback: string): string {
  for (const k of keys) {
    const v = params[k];
    if (typeof v === "string" && v.trim().length > 0) return v.trim();
  }
  return fallback;
}

function extractSavedEvents(rows: unknown[]): EventRow[] {
  const out: EventRow[] = [];
  for (const row of rows) {
    if (!row || typeof row !== "object") continue;
    const r = row as AnyObj;
    if (r.phase !== "accepted_extrema") continue;
    const side = r.side;
    const kind = r.kind;
    if ((side !== "yes" && side !== "no") || (kind !== "valley" && kind !== "peak")) continue;
    const ts = fnum(r.emitted_ts_ms) ?? fnum(r.ts_ms);
    if (ts == null) continue;
    out.push({ ts_ms: Math.trunc(ts), side, kind });
  }
  out.sort((a, b) => a.ts_ms - b.ts_ms || a.side.localeCompare(b.side) || a.kind.localeCompare(b.kind));
  return out;
}

function roundStart(tsMs: number): number {
  return tsMs - (tsMs % WINDOW_MS);
}

function byWindow(events: EventRow[], startMs: number | null, rounds: number): EventRow[] {
  if (startMs == null) return events;
  const endMs = startMs + rounds * WINDOW_MS;
  return events.filter((x) => x.ts_ms >= startMs && x.ts_ms < endMs);
}

function sideEventsForRound(
  points: AnyObj[],
  side: "yes" | "no",
  params: AnyObj,
): EventRow[] {
  const askKey = side === "yes" ? "yes_ask" : "no_ask";
  const bidKey = side === "yes" ? "yes_bid" : "no_bid";
  const midKey = side;

  const times: number[] = [];
  const askRaw: Array<number | null> = [];
  const bidRaw: Array<number | null> = [];
  for (const p of points) {
    const ts = fnum(p.timestamp_ms);
    if (ts == null) continue;
    const ask = fnum(p[askKey]) ?? fnum(p[midKey]);
    const bid = fnum(p[bidKey]) ?? ask;
    if (ask == null || bid == null) continue;
    times.push(Math.trunc(ts));
    askRaw.push(ask);
    bidRaw.push(bid);
  }
  if (times.length === 0) return [];

  const span = Math.trunc(pnum(params, ["span"], 20));
  const noiseFilter = pstr(params, ["noise_filter", "noiseFilter"], "gaussian") as
    | Parameters<typeof applyNoiseFilter>[2]
    | string;
  const quoteHyst = pnum(
    params,
    ["quote_structural_hysteresis_cents", "quoteStructuralHysteresisCents"],
    0.1,
  );
  const sustainedMs = Math.trunc(pnum(params, ["sustained_tangent_ms", "sustainedTangentMs"], 1000));
  const valleyAskMax = pnum(params, ["valley_max_raw_ask_cents", "valleyMaxRawAskCents"], 45);
  const v2pMin = pnum(
    params,
    ["valley_to_peak_raw_spread_min_cents", "valleyToPeakRawSpreadMinCents"],
    10,
  );
  const locationMode = pstr(params, ["extrema_location_mode", "extremaLocationMode"], "spread_anchored")
    .trim()
    .toLowerCase();

  const smooth = applyNoiseFilter(askRaw, span, noiseFilter as Parameters<typeof applyNoiseFilter>[2]);
  const raw = discreteTangentExtremaIndices(smooth, times, {
    structuralHysteresis: quoteHyst,
    sustainedTangentMs: sustainedMs,
    cooldownMs: 1000,
    goldenPeakFallbackWindowMs: 30_000,
    maxValleyAskExclusive: valleyAskMax,
    valleyAskSeries: askRaw,
    peakBidSeries: bidRaw,
    anchorToQuoteSeries: locationMode === "spread_anchored",
  });
  const accepted = filterAlternatingAskSwingCents(
    raw.peaks,
    raw.valleys,
    askRaw,
    bidRaw,
    v2pMin,
    0,
    times,
    1000,
  );
  const evs: EventRow[] = [];
  for (const i of accepted.valleys) evs.push({ ts_ms: times[i]!, side, kind: "valley" });
  for (const i of accepted.peaks) evs.push({ ts_ms: times[i]!, side, kind: "peak" });
  evs.sort((a, b) => a.ts_ms - b.ts_ms || a.side.localeCompare(b.side) || a.kind.localeCompare(b.kind));
  return evs;
}

function recomputeTsEvents(result: AnyObj): EventRow[] {
  const meta = (result.meta as AnyObj | undefined) ?? {};
  const params = (meta.strategy_params as AnyObj | undefined) ?? {};
  const equityRaw = result.equity;
  if (!Array.isArray(equityRaw)) throw new Error("Result JSON has no `equity` array");
  const points = equityRaw.filter((x): x is AnyObj => !!x && typeof x === "object");
  points.sort((a, b) => (fnum(a.timestamp_ms) ?? -1) - (fnum(b.timestamp_ms) ?? -1));

  const rounds = new Map<number, AnyObj[]>();
  for (const p of points) {
    const ts = fnum(p.timestamp_ms);
    if (ts == null) continue;
    const rs = roundStart(Math.trunc(ts));
    const arr = rounds.get(rs) ?? [];
    arr.push(p);
    rounds.set(rs, arr);
  }

  const all: EventRow[] = [];
  for (const [, roundPts] of [...rounds.entries()].sort((a, b) => a[0] - b[0])) {
    all.push(...sideEventsForRound(roundPts, "yes", params));
    all.push(...sideEventsForRound(roundPts, "no", params));
  }
  all.sort((a, b) => a.ts_ms - b.ts_ms || a.side.localeCompare(b.side) || a.kind.localeCompare(b.kind));
  return all;
}

function compare(saved: EventRow[], recomputed: EventRow[]): number {
  const n = Math.min(saved.length, recomputed.length);
  for (let i = 0; i < n; i++) {
    const a = saved[i]!;
    const b = recomputed[i]!;
    if (a.ts_ms !== b.ts_ms || a.side !== b.side || a.kind !== b.kind) {
      console.log("Mismatch at index", i);
      console.log(" saved     ", a);
      console.log(" ts-side   ", b);
      return 1;
    }
  }
  if (saved.length !== recomputed.length) {
    console.log("Length mismatch", { saved: saved.length, ts_side: recomputed.length });
    if (saved.length > n) console.log(" next saved   ", saved[n]);
    if (recomputed.length > n) console.log(" next ts-side ", recomputed[n]);
    return 1;
  }
  console.log("OK: saved debug events and TS recompute match exactly.");
  console.log("event_count =", saved.length);
  return 0;
}

function main(): number {
  const { jsonPath, roundStartMs, rounds } = parseArgs(process.argv);
  const full = path.resolve(process.cwd(), jsonPath);
  const obj = JSON.parse(fs.readFileSync(full, "utf-8")) as AnyObj;
  const debugEvents = obj.strategy_debug_events;
  if (!Array.isArray(debugEvents)) {
    throw new Error("No `strategy_debug_events` in result JSON. Re-run mike_v1 with `debug_trace=true`.");
  }
  const saved = byWindow(extractSavedEvents(debugEvents), roundStartMs, rounds);
  const recomputed = byWindow(recomputeTsEvents(obj), roundStartMs, rounds);
  return compare(saved, recomputed);
}

process.exitCode = main();
